### PR TITLE
Generalize handler APIs for return nullability on (un-)annotated code

### PR DESCRIPTION
**Pull request by lazaroclapp**
This is a follow up to changes in #639, #636, and #624. Here we
generalize the former Handler callback:

- onUnannotatedInvocationGetExplicitlyNonNullReturn

which was used to change the nullability of method returns in
unannotated code for the purpose of method `@Override` checks
(i.e. subtyping).
    
We provide instead:
    
- onOverrideMethodInvocationReturnNullability
    
Which allows handlers to change the nullability of method
returns for both annotated and unannotated code, for the same
purposes.
    
Additionally, we implement this new handler callback in
`LibraryModelsHandler`.
    
Before this change, marking a return as non-null in a library model
affected checking of method invocations, but not necessarily
overriding/subtyping checks.

### CODE REVIEW

File: guava-recent-unit-tests/src/test/java/com/uber/nullaway/guava/NullAwayGuavaParametricNullnessTests.java
Comment (msridhar) [inline guava-recent-unit-tests/src/test/java/com/uber/nullaway/guava/NullAwayGuavaParametricNullnessTests.java:130]: Why these new imports?

File: nullaway/src/test/java/com/uber/nullaway/NullAwayAcknowledgeRestrictiveAnnotationsTests.java
Comment (msridhar) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayAcknowledgeRestrictiveAnnotationsTests.java:303]:
```suggestion
                // Note: this is the OFF case.
```

**Review by msridhar — APPROVED** [review]
Mostly LGTM!  Approved assuming comment and nit are addressed

**Comment by msridhar** [comment]
Actually, @lazaroclapp did we still want to do some perf testing here before landing?

File: guava-recent-unit-tests/src/test/java/com/uber/nullaway/guava/NullAwayGuavaParametricNullnessTests.java
Comment (lazaroclapp) [reply guava-recent-unit-tests/src/test/java/com/uber/nullaway/guava/NullAwayGuavaParametricNullnessTests.java:130]: Removing. Left-over from when this test code was all in the same test.

**Comment by lazaroclapp** [comment]
> Actually, @lazaroclapp did we still want to do some perf testing here before landing?

I think that would be a good idea. Though, ideally, we want to test this PR vs the current NullAway release, not current main. So not sure if that's easier before or after landing...

**Comment by msridhar** [comment]
> > Actually, @lazaroclapp did we still want to do some perf testing here before landing?
> 
> I think that would be a good idea. Though, ideally, we want to test this PR vs the current NullAway release, not current main. So not sure if that's easier before or after landing...

I should be able to test before landing.  Will set it up now.  I will test vs. the version after #631 landed, as that PR gave a speed improvement.

**Comment by msridhar** [comment]
Initial perf comparison of this PR vs. right after #631 landed (commit c3711177ceffabdb35aa06bfaf9a90b6a8d49100).

After #631:

```
Benchmark                          Mode  Cnt   Score   Error  Units
AutodisposeBenchmark.compile      thrpt   25  10.739 ± 0.089  ops/s
CaffeineBenchmark.compile         thrpt   25   2.517 ± 0.013  ops/s
DFlowMicroBenchmark.compile       thrpt   25  28.917 ± 0.149  ops/s
NullawayReleaseBenchmark.compile  thrpt   25   1.509 ± 0.013  ops/s
```

After this PR:

```
Benchmark                          Mode  Cnt   Score   Error  Units
AutodisposeBenchmark.compile      thrpt   25  10.605 ± 0.078  ops/s
CaffeineBenchmark.compile         thrpt   25   2.407 ± 0.014  ops/s
DFlowMicroBenchmark.compile       thrpt   25  28.158 ± 0.203  ops/s
NullawayReleaseBenchmark.compile  thrpt   25   1.492 ± 0.011  ops/s
```

The regression on Caffeine (about 4.5%) is a bit concerning.  Let me try a couple more rounds of benchmarking with Caffeine to try to narrow things down a bit more.

**Comment by lazaroclapp** [comment]
We don't get any profiling data out of those benchmarks, right? Would be interesting to see if `onOverrideMethodInvocationReturnNullability(...)` specifically is a hot method for `CaffeineBenchmark`...

**Comment by msridhar** [comment]
After further investigation I've narrowed the perf regression down to #639.  Here are the numbers.

Before #639 (48e573965a48df957e98b003788bbab328511cc1):
```
Benchmark                   Mode  Cnt  Score   Error  Units
CaffeineBenchmark.compile  thrpt   25  2.503 ± 0.013  ops/s
```

After #639 (0e575ce8e0d8fa7f07404ba04c30c004348ec50b):
```
Benchmark                   Mode  Cnt  Score   Error  Units
CaffeineBenchmark.compile  thrpt   25  2.416 ± 0.007  ops/s
```

I don't believe this PR introduces any additional regression beyond that of #639.

@lazaroclapp let me know how you want to proceed.  Maybe we can land this PR and also open an issue on fixing the #639 regression?

**Comment by lazaroclapp** [comment]
> Maybe we can land this PR and also open an issue on fixing the https://github.com/uber/NullAway/pull/639 regression?

Sounds like a plan to me! Do you have any ideas on how to narrow down / profile the cause of the regression? (Would be good if we can resolve before NullAway 0.9.10, though not critical, I think)

**Comment by msridhar** [comment]
> > Maybe we can land this PR and also open an issue on fixing the #639 regression?
> 
> Sounds like a plan to me! Do you have any ideas on how to narrow down / profile the cause of the regression? (Would be good if we can resolve before NullAway 0.9.10, though not critical, I think)

I think we just need to compare profiles before and after, and hopefully the change will jump out.  I'll try to take a quick look soon.

**Pull request merged**
